### PR TITLE
Ready: CRDT API Improvements

### PR DESCRIPTION
**Pull request by lukebakken**
CLIENTS-370

### CODE REVIEW

**Comment by alexmoore** [comment]
One thing that bugs me is the massiveness of this namespace:

windows_8_1_x64

Could we break that up into either Operation folders (CRDT/Set, CRDT/Map), or associated types (CRDT/Builders, CRDT/Response) ?

File: src/RiakClient/Commands/CRDT/FetchCommandBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+{
+    using System;
+
+    /// <summary>
+    /// Builds a fetch command.
+    /// </summary>
+    /// <typeparam name="TCommand">The type of the fetch command.</typeparam>
+    /// <typeparam name="TOptions">The type of the options for the fetch command.</typeparam>
+    /// <typeparam name="TResponse">The type of the fetch command's response.</typeparam>
+    public abstract class FetchCommandBuilder<TCommand, TOptions, TResponse>
```
Comment (alexmoore) [inline src/RiakClient/Commands/CRDT/FetchCommandBuilder.cs]: Niiiiice.

**Comment by lukebakken** [comment]
@alexmoore more namespaces = more `using` statements for end-users. Also, note that as we move entirely towards `IRiakCommand` for all operations all of the abstract classes in this namespace will be used for all commands, not just CRDT ones.

**Comment by alexmoore** [comment]
@lukebakken Ok. Let's take this as the maximum for files in a namespace then.

**Comment by lukebakken** [comment]
@alexmoore I don't really care so much about namespace size as much as ensuring that related classes and types are all in the same namespace. Other than that, I don't really see the use of breaking stuff up...

Example one of many from the mothership itself:
https://github.com/aspnet/Mvc/tree/dev/src/Microsoft.AspNet.Mvc.Core

File: src/RiakClient/Comms/RiakConnection.cs [hunk 2]
```diff
@@ -290,6 +302,29 @@ public void Disconnect()
...
+                return socketFunc();
+            }
+            catch (RiakException ex)
+            {
+                if (ex.NodeOffline)
+                {
+                    Disconnect();
+                }
+
+                // TODO: we really should preserve the Exception object
```
Comment (alexmoore) [inline src/RiakClient/Comms/RiakConnection.cs]: Make it so.

File: src/RiakClient/IRiakBatchClient.cs [hunk 3]
```diff
@@ -323,7 +323,7 @@ public interface IRiakBatchClient
         /// <b>Callers should fully enumerate the collection or else close the connection when finished.</b>
         /// </remarks>
         RiakResult<IEnumerable<string>> StreamListBuckets();
-
```
Comment (alexmoore) [inline src/RiakClient/IRiakBatchClient.cs]: Noooo 
image

File: src/RiakClient/Messages/riak_yokozuna.cs [hunk 4]
```diff
@@ -1,23 +1,13 @@
...
-// KIND, either express or implied.  See the License for the
-// specific language governing permissions and limitations
-// under the License.
-// </copyright>
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
 
-#pragma warning disable 1591
```
Comment (alexmoore) [inline src/RiakClient/Messages/riak_yokozuna.cs]: Did we disable 1591 project wide?

File: src/RiakClient/Models/Index/SecondaryIndex.cs [hunk 5]
```diff
@@ -30,7 +30,6 @@ namespace RiakClient.Models.Index
     /// </summary>
     /// <typeparam name="TClass">The concrete type of the implementing class.</typeparam>
     /// <typeparam name="TIndex">The type of the index (<see cref="BigInteger"/> or <see cref="string"/>).</typeparam>
-    [ComVisible(false)]
```
Comment (alexmoore) [inline src/RiakClient/Models/Index/SecondaryIndex.cs]: Also what did we change that unclogged a bunch of these not com visible classes?

File: src/RiakClient/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -24,11 +24,12 @@
 
 [assembly: AssemblyTitle("Riak .NET Client")]
 [assembly: AssemblyDescription(".NET Client for the Riak distributed database.")]
-[assembly: ComVisible(true)]
+[assembly: ComVisible(false)]
```
Comment (alexmoore) [inline src/RiakClient/Properties/AssemblyInfo.cs]: Ahh, nevermind.

File: src/RiakClient/Messages/riak.cs [hunk 7]
```diff
@@ -504,6 +494,14 @@ public bool consistent
       get { return _consistent; }
       set { _consistent = value; }
     }
+    private bool _write_once = default(bool);
+    [global::ProtoBuf.ProtoMember(28, IsRequired = false, Name=@"write_once", DataFormat = global::ProtoBuf.DataFormat.Default)]
+    [global::System.ComponentModel.DefaultValue(default(bool))]
+    public bool write_once
```
Comment (alexmoore) [inline src/RiakClient/Messages/riak.cs]: Remember to add this to bucket properties.

**Comment by alexmoore** [comment]
+1 as long as we note in JIRA to add the bucket property, new yokozuna option later on.

File: src/RiakClient/Messages/riak_yokozuna.cs (hunk 4 above)
Comment (lukebakken) [inline src/RiakClient/Messages/riak_yokozuna.cs]: Yeah apparently a while ago. I'll get rid of the other pragmas for it.

File: src/RiakClient/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (lukebakken) [inline src/RiakClient/Properties/AssemblyInfo.cs]: NO COM FOR YOU

Comment (alexmoore) [inline src/RiakClient/Properties/AssemblyInfo.cs]: Noooo, but think of the poor MFC projects!

**Comment by alexmoore** [comment]
+1, we have those tickets in JIRA.

**Comment by lukebakken** [comment]
Roger that.

**Pull request merged**